core/src/main/java/org/apache/calcite/adapter/enumerable/RexImpTable.java

```diff
@@ -678,7 +682,13 @@ Builder populate2() {
       final RexCallImplementor value = new ValueConstructorImplementor();
       map.put(MAP_VALUE_CONSTRUCTOR, value);
       map.put(ARRAY_VALUE_CONSTRUCTOR, value);
-      map.put(ITEM, new ItemImplementor());
+
+      // Each subscript operator has a corresponding offset and 'safe' value.
```

Comment:
Better to move `offset` and `safe` from `ItemImplementor` into `SqlItemOperator` (make them public final fields)

Right now you're leaving a little too heavily on object identity (and there's a danger that future maintainers will branch based on the name of the operator, which will be bad).